tests/unit/kernel/MaternKernel_unittest.cc

```diff
@@ -0,0 +1,142 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+#include <gtest/gtest.h>
+#include <shogun/kernel/MaternKernel.h>
+#include <shogun/lib/SGMatrix.h>
+#include <shogun/util/zip_iterator.h>
+
+using namespace shogun;
+
+TEST(MaternKernelTest_1_2, test_kernel_matrix)
+{
+	SGMatrix<float64_t> matrix{{0, 1, 2}, {3, 4, 5}};
+	SGMatrix<float64_t> expected_kernel_matrix{{1., 0.07441661},
+											   {0.07441661, 1.}};
+
+	auto features = std::make_shared<DenseFeatures<float64_t>>(matrix);
+
+	auto kernel = std::make_shared<MaternKernel>(features, features, 10, 2.0, 0.5);
+	SGMatrix<float64_t> computed_kernel_matrix = kernel->get_kernel_matrix();
+
+	for (const auto& [result, reference]: zip_iterator(computed_kernel_matrix, expected_kernel_matrix))
+		EXPECT_NEAR(result, reference, 1E-7);		
+}
+
+TEST(MaternKernelTest_3_2, test_kernel_matrix)
+{
+	SGMatrix<float64_t> matrix{{0, 1, 2}, {3, 4, 5}};
+	SGMatrix<float64_t> expected_kernel_matrix{{1., 0.06109951},
+											   {0.06109951, 1.}};
+
+	auto features = std::make_shared<DenseFeatures<float64_t>>(matrix);
+
+	auto kernel = std::make_shared<MaternKernel>(features, features, 10, 2.0, 1.5);
+	SGMatrix<float64_t> computed_kernel_matrix = kernel->get_kernel_matrix();
+
+	for (const auto& [result, reference]: zip_iterator(computed_kernel_matrix, expected_kernel_matrix))
+		EXPECT_NEAR(result, reference, 1E-7);		
+}
+
+TEST(MaternKernelTest_5_2, test_kernel_matrix)
+{
+	SGMatrix<float64_t> matrix{{0, 1, 2}, {3, 4, 5}};
+	SGMatrix<float64_t> expected_kernel_matrix{{1., 0.05416044},
+											   {0.05416044, 1.}};
+
+	auto features = std::make_shared<DenseFeatures<float64_t>>(matrix);
+
+	auto kernel = std::make_shared<MaternKernel>(features, features, 10, 2.0, 2.5);
+	SGMatrix<float64_t> computed_kernel_matrix = kernel->get_kernel_matrix();
+
+	for (const auto& [result, reference]: zip_iterator(computed_kernel_matrix, expected_kernel_matrix))
+		EXPECT_NEAR(result, reference, 1E-7);		
+}
+
+TEST(MaternKernelTest_1_2, test_derivative_width)
+{
+	// gradient computed with JAX: https://gist.github.com/gf712/b10ecefd2379e8492c9802ac8cb329c9
+	SGMatrix<float64_t> matrix{{0, 1, 2}, {3, 4, 5}};
+	SGMatrix<float64_t> expected_dmatrix{{0., 0.09667},
+										 {0.09667, 0.}};
+
+	auto features = std::make_shared<DenseFeatures<float64_t>>(matrix);
+
+	auto kernel = std::make_shared<MaternKernel>(features, features, 10, 2.0, 0.5);
+	auto params = kernel->get_params();
+	auto width = params.find("width");
+	SGMatrix<float64_t> dMatrix = kernel->get_parameter_gradient(*width);
+
+	for (const auto& [result, reference]: zip_iterator(dMatrix, expected_dmatrix))
+		EXPECT_NEAR(result, reference, 1E-7);		
```

Comment:
single precision?

Reply:
You mean get epsilon from type traits?

Reply:
If that is possible I think it should be done or?

Reply:
I am not sure, because the error propagates through each calculation. This is the tolerance value that got everything correct for all the tests without tweaking for each fixture